source/slang/slang-check-stmt.cpp

```diff
@@ -239,6 +239,48 @@ namespace Slang
         subContext.checkStmt(stmt->body);
     }
 
+    void SemanticsStmtVisitor::validateCaseStmts(SwitchStmt* stmt, DiagnosticSink* sink)
+    {
+        auto blockStmt = as<BlockStmt>(stmt->body);
+        if (!blockStmt)
+            return;
+
+        auto seqStmt = as<SeqStmt>(blockStmt->body);
+        if (!seqStmt)
+            return;
+
+        bool hasDefaultStmt = false;
+        HashSet<Val*> caseStmtVals;
+        for (auto& sStmt : seqStmt->stmts)
+        {
+            // check that all case tags are unique
+            if (auto caseStmt = as<CaseStmt>(sStmt))
+            {
+                if (caseStmt->exprVal)
+                {
+                    // exprVal contains the constant folded expr, that is checked for
+                    // uniqueness within the scope of the switch statement.
+                    if (!caseStmtVals.add(caseStmt->exprVal))
+                    {
+                        sink->diagnose(sStmt, Diagnostics::duplicateCasesInASwitch);
+                        return;
```

Comment:
I am just thinking loud here.
I am wondering if it is possible for two different memory pointers to point to a same value.

Reply:
In our ASTs, everything inherits from `Val` is globally deduplicated, meaning that Vals will be pointer-unique.
The same applies for all IRInst* that are `HOISTABLE`.

Reply:
thanks for the comment. I added some more explanation where the const folding is done.